ExtendedProtectionPolicy constructor throws unnecessarily on WASM

### Description

This was an issue discovered by a WCF customer. WCF tracking issue is dotnet/wcf#4942.  

It's not possible to construct an instance of `ExtendedProtectionPolicy` on WASM. This makes it impossible to have a public API expose this type unless you return null when running in a browser. `ExtendedProtectionPolicy` has the ability to have a policy enforcement that it is never used, which is appropriate for platforms where extended protection isn't supported.  

In addition to this problem, the property `ExtendedProtectionPolicy.OSSupportsExtendedProtection` always returns true with no regard to platform.  

WCF exposes a property of type `ExtendedProtectionPolicy` to enable clients to optionally turn on the use of Extended Protection. The default value for this property is a policy where enforcement is disabled. This is constructed using [this line of code](https://github.com/dotnet/wcf/blob/fd1bd239d29a8de6bd4b7829460b98202a657c14/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ChannelBindingUtility.cs#L14):
```c#
        private static readonly ExtendedProtectionPolicy s_disabledPolicy = new ExtendedProtectionPolicy(PolicyEnforcement.Never);
```
In our property, [when a new value is being set](https://github.com/dotnet/wcf/blob/fd1bd239d29a8de6bd4b7829460b98202a657c14/src/System.Private.ServiceModel/src/System/ServiceModel/HttpTransportSecurity.cs#L73-L78), we check if the new value has an enforcement policy set to `Always`, and if it is we check the value of `ExtendedProtectionPolicy.OSSupportsExtendedProtection` and throw if trying to set a policy which can't be supported.  
```c#
                if (value.PolicyEnforcement == PolicyEnforcement.Always &&
                    !System.Security.Authentication.ExtendedProtection.ExtendedProtectionPolicy.OSSupportsExtendedProtection)
                {
                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
                        new PlatformNotSupportedException(SR.ExtendedProtectionNotSupported));
                }
```
If `OSSupportsExtendedProtection` were implemented to reflect whether extended protection, then this should be sufficient for apps to guard against trying to use extended protection where it's not supported.  

Throwing an exception on construction of a policy instance regardless of enforcement policy means all public properties need to return null where previously they didn't. This changes a not null value to a sometimes null value. This also requires adding logic to interpret null to mean the same as `PolicyEnforcement.Never` and duplicating that logic change everywhere the instance propagates through the code base.

### Reproduction Steps

In a Blazor app, create a new instance of `BasicHttpBinding` using the default constructor. It throws with a `PlatformNotSupportedException` stating "System.Net.Security is not supported on this platform."  

This was customer reported, but based on my understanding of how the throwing of an exception is generated, I suspect getting `ExtendedProtectionPolicy.OSSupportsExtendedProtection` will throw a `PlatformNotSupportedException` too instead of returning false.

### Expected behavior

When constructing an instance with a `PolicyEnforcement` value of `Never` or `WhenSupported` will not throw. These classes are just OM around describing extended protection so I would expect an exception to only be thrown when a class which accepts an `ExtendedProtectionPolicy` is passed one with `PolicyEnforcement.Always`, as WCF does. In this case, the class would never throw. 
  
This isn't a new expectation; this is what happened on OS's earlier than Win7 where extended protection isn't supported.

### Actual behavior

Constructing an `ExtendedProtectionPolicy` with any enforcement policy throws a PlatformNotSupportedException.

### Regression?

Regression from .NET Framework and .NET Core outside of a browser. On .NET Framework, OS's prior to Windows 7 didn't support extended protection, and `ExtendedProtectionPolicy.OSSupportsExtendedProtection` would return false so that libraries and applications can avoid an exception being thrown by trying to use extended protection when it's not supported. The same model should work for .NET too.

### Known Workarounds

Add lots of ugly logic through the WCF code base treating null as `PolicyEnforcement.Never`. This could also cause consuming code to throw a `NullReferenceException` if expecting the property to return a non-null value as it does on every other platform.

### Configuration

_No response_

### Other information

_No response_

IIUC, for:
https://github.com/dotnet/runtime/blob/076f8c58984b009033204ad1dd9f86c691ae42fb/src/libraries/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicy.cs#L147-L154

.. we want to have this property return `false` on `wasm.

But `System.Net.Security` has:

https://github.com/dotnet/runtime/blob/076f8c58984b009033204ad1dd9f86c691ae42fb/src/libraries/System.Net.Security/Directory.Build.props#L6

.. which would cause the assembly to have auto-generated PNSEs.

What would be a good way to have this property return `false` for wasm?

1. I tried a substitution xml, but that doesn't seem to take affect. The assembly still gets the PNSE.
2. I tried to include a partial class with just the property for wasm, but that fails with:
`obj/System.Net.Security/Release/net8.0/System.Net.Security.notsupported.cs(710,28): error CS0102: The type 'ExtendedProtectionPolicy' already contains a definition for 'OSSupportsExtendedProtection'`

cc @eerhardt

--- a/src/libraries/System.Net.Security/src/ExcludeApiList.PNSE.txt
+++ b/src/libraries/System.Net.Security/src/ExcludeApiList.PNSE.txt
@@ -1,0 +1,2 @@
+System.Security.Authentication.ExtendedProtection.ExtendedProtectionPolicy
+System.Security.Authentication.ExtendedProtection.ServiceNameCollection

--- a/src/libraries/System.Net.Security/src/System.Net.Security.csproj
+++ b/src/libraries/System.Net.Security/src/System.Net.Security.csproj
@@ -9,6 +9,7 @@
   <PropertyGroup>
     <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
     <GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetPlatformIdentifier)' == ''">SR.SystemNetSecurity_PlatformNotSupported</GeneratePlatformNotSupportedAssemblyMessage>
+    <ApiExclusionListPath Condition="'$(TargetPlatformIdentifier)' == ''">ExcludeApiList.PNSE.txt</ApiExclusionListPath>
     <DefineConstants Condition="'$(TargetPlatformIdentifier)' == 'windows'">$(DefineConstants);TARGET_WINDOWS</DefineConstants>
     <DefineConstants Condition="'$(TargetPlatformIdentifier)' == 'android'">$(DefineConstants);TARGET_ANDROID</DefineConstants>
     <UseAndroidCrypto Condition="'$(TargetPlatformIdentifier)' == 'android'">true</UseAndroidCrypto>
@@ -18,6 +19,14 @@
     <GenAPIExcludeApiList>ReferenceAssemblyExclusions.txt</GenAPIExcludeApiList>
   </PropertyGroup>
   <Import Project="$(CommonPath)System\Security\Cryptography\Asn1Reader\System.Security.Cryptography.Asn1Reader.Shared.projitems" Condition="'$(UseManagedNtlm)' == 'true'" />
+  <ItemGroup>
+    <Compile Include="System\Security\Authentication\ExtendedProtection\ExtendedProtectionPolicy.cs" />
+    <Compile Include="System\Security\Authentication\ExtendedProtection\ServiceNameCollection.cs" />
+    <Compile Include="$(CommonPath)System\Obsoletions.cs"
+             Link="Common\System\Obsoletions.cs" />
+    <Compile Include="$(CommonPath)System\Net\UriScheme.cs"
+             Link="Common\System\Net\UriScheme.cs" />
+  </ItemGroup>
   <ItemGroup Condition="'$(TargetPlatformIdentifier)' != ''">
     <Compile Include="System\Net\CertificateValidationPal.cs" />
     <Compile Include="System\Net\SslStreamContext.cs" />
@@ -53,10 +62,8 @@
     <Compile Include="System\Net\NTAuthentication.cs" />
     <Compile Include="System\Net\StreamFramer.cs" />
     <Compile Include="System\Net\Security\NegotiateStream.cs" />
-    <Compile Include="System\Security\Authentication\ExtendedProtection\ExtendedProtectionPolicy.cs" />
     <Compile Include="System\Security\Authentication\ExtendedProtection\PolicyEnforcement.cs" />
     <Compile Include="System\Security\Authentication\ExtendedProtection\ProtectionScenario.cs" />
-    <Compile Include="System\Security\Authentication\ExtendedProtection\ServiceNameCollection.cs" />
     <!-- IP parser -->
     <Compile Include="$(CommonPath)System\Net\IPv4AddressHelper.Common.cs"
              Link="System\Net\IPv4AddressHelper.Common.cs" />
@@ -87,8 +94,6 @@
              Link="Common\System\Net\ExceptionCheck.cs" />
     <Compile Include="$(CommonPath)System\Net\SecurityProtocol.cs"
              Link="Common\System\Net\SecurityProtocol.cs" />
-    <Compile Include="$(CommonPath)System\Net\UriScheme.cs"
-             Link="Common\System\Net\UriScheme.cs" />
     <!-- Common -->
     <Compile Include="$(CommonPath)System\NotImplemented.cs"
              Link="Common\System\NotImplemented.cs" />
@@ -106,8 +111,6 @@
              Condition="'$(UseManagedNtlm)' != 'true'" />
     <Compile Include="$(CommonPath)System\HexConverter.cs"
              Link="Common\System\HexConverter.cs" />
-    <Compile Include="$(CommonPath)System\Obsoletions.cs"
-             Link="Common\System\Obsoletions.cs" />
     <Compile Include="$(CommonPath)Interop\Windows\SChannel\Interop.SECURITY_STATUS.cs"
              Link="Common\Interop\Windows\SChannel\Interop.SECURITY_STATUS.cs" />
   </ItemGroup>

--- a/src/libraries/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicy.cs
+++ b/src/libraries/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicy.cs
@@ -149,7 +149,7 @@ namespace System.Security.Authentication.ExtendedProtection
             get
             {
                 // .NET Core is supported only on Win7+ where ExtendedProtection is supported.
-                return true;
+                return OperatingSystem.IsWindows();
             }
         }
     }

--- a/src/libraries/System.Net.Security/tests/UnitTests/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicyTest.cs
+++ b/src/libraries/System.Net.Security/tests/UnitTests/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicyTest.cs
@@ -57,9 +57,17 @@ namespace System.Net.Security.Tests
         }
 
         [Fact]
-        public void ExtendedProtectionPolicy_OSSupportsExtendedProtection()
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void ExtendedProtectionPolicy_OSSupportsExtendedProtection_Windows()
         {
             Assert.True(ExtendedProtectionPolicy.OSSupportsExtendedProtection);
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        public void ExtendedProtectionPolicy_OSSupportsExtendedProtection_NonWindows()
+        {
+            Assert.False(ExtendedProtectionPolicy.OSSupportsExtendedProtection);
         }
 
         [Fact]

